skorch/_doctor.py

```diff
@@ -0,0 +1,936 @@
+            if not isinstance(v, np.ndarray):
+                raise TypeError(f"Activations of type {type(v)} are not supported")
+            logs[-1][layer_name + f'["{k}"]'] = v
+    else:
+        raise TypeError(f"Activations of type {type(v)} are not supported")
+
+
+def add_activation_hooks(net):
+    """Add forward hooks to all layers to record activations
+
+    Ignore the top level modules like ``net.module_`` and ``net.criterion_``.
+
+    If an output is not a simple array, it is disambiguated. E.g. if it's a
+    list, the name get a suffix of ``[i]`` where ``i`` designates the index in
+    the list. Similary, when the output is a dict, a ``[key]`` suffix is added,
+    where ``[key]`` is the key of the corresponding value in the dictionary.
+
+    Parameters
+    ----------
+    net : skorch.NeuralNet
+      The neural net instance (has to be initialized).
+
+    Returns
+    -------
+    logs : dict of list of dict
+      Data structure containing the logged activations. For each module, for
+      each training step, for each layer, there is an entry of the recorded
+      activations.
+
+    handles : list of torch.utils.hooks.RemovableHandle
+      These handles are returned by torch when adding hooks. They can be used to
+      remove the hooks.
+
+    Raises
+    ------
+    TypeError
+      When the output of a layer is not a simple array and if it cannot be
+      disambiguated, a ``TypeError`` is raised. Disambiguation works for lists,
+      tuples, and dicts (but not nested ones).
+
+    """
+    logs = {}
+    handles = []
+
+    for module_name, module in named_modules(net):
+        logs[module_name] = []
+        for layer_name, submodule in module.named_modules():
+            if submodule is module:
+                # is logging activations for whole module useful?
+                continue
+
+            handle = submodule.register_forward_hook(partial(
+                _add_activation_hook,
+                logs=logs[module_name],
+                layer_name=layer_name,
+            ))
+            handles.append(handle)
+
+    return logs, handles
+
+
+def _add_grad_hook(grad, *, log_grad, log_param_update, param_name, tensor):
+    """Helper function for adding gradient hooks"""
+    # if the param is already there, the entry is from the previous training
+    # step, thus start a new one
+
+    # very first batch
+    if not log_grad:
+        log_grad.append({})
+    if not log_param_update:
+        log_param_update.append({})
+
+    if param_name in log_grad[-1]:
+        log_grad.append({})
+    if param_name in log_param_update[-1]:
+        log_param_update.append({})
+
+    # We need to clone the grad here because otherwise, the recorded gradients
+    # can be overridden by later gradients despite them being pulled to numpy!
+    # This only happens on CPU, which suggests it's a strange bug that's perhaps
+    # related to caching and/or memory-mapping.
+    grad = grad.clone()
```

Comment:
For the gradient, I do not think it matters too much, but I think `detach` is better:

```suggestion
    grad = grad.detach()
```